### PR TITLE
[biomed nl] multi entity fulfill

### PR DESCRIPTION
**Pull request by chejennifer**
NL changes for fulfilling multiple entities

frontend changes in: https://github.com/datacommonsorg/website/pull/4070

Screenshot 2024-03-26 at 7 10 49 PM

### CODE REVIEW

File: server/lib/nl/fulfillment/triple.py [hunk 1]
```diff
@@ -32,17 +32,34 @@
...
 _IN_TITLE = '{entity} is the {property} for:'
+_MULTI_ENTITY_TITLE = 'The {property} for {entity} are as follows:'
+_MAX_ENTITIES_IN_TITLE = 2
+
+
+# Gets a title string for a list of entities
+def _get_entity_string(entities: List[Entity]) -> str:
+  if len(entities) == 1:
+    return entities[0].name or entities[0].dcid
+  entities_cutoff_idx = min(_MAX_ENTITIES_IN_TITLE, len(entities) - 1)
```
[the remark this answers is not in the document]
Comment (pradh) [reply server/lib/nl/fulfillment/triple.py:43]: I wonder if the logic would be simpler as a single for loop.

```
max_entries = min(3, len(entities))
for i, e in enumerate(entities[:max_entries]):
  if i > 0:
    # Delim only for 2nd entry onwards
    if i == max_entries - 1:
      parts.append('and ')
    else:
      parts.append(', ')
  parts.append(e.name or e.dcid)
return ''.join(parts)
```

**Review by pradh — COMMENTED** [review]
Nice!

File: server/lib/nl/fulfillment/triple.py [hunk 2]
```diff
@@ -32,17 +32,34 @@
...
+      [e.name or e.dcid for e in entities[:entities_cutoff_idx]])
+  entity_str += f' and {entities[entities_cutoff_idx].name or entities[entities_cutoff_idx].dcid}'
+  return entity_str
 
 
 # Gets the chart title for a list of entities and a property expression
 # TODO: revisit how titles should be displayed
 def _get_title(entities: List[Entity], prop_expression: str) -> str:
-  entity_str = ', '.join([e.name or e.dcid for e in entities])
+  entity_str = _get_entity_string(entities)
```
[the remark this answers is not in the document]
Comment (pradh) [reply server/lib/nl/fulfillment/triple.py:53]: If you agree a TODO is fine.

**Review by pradh — APPROVED** [review]

File: server/lib/nl/fulfillment/triple.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chejennifer) [reply server/lib/nl/fulfillment/triple.py:67]: added TODO

File: server/lib/nl/fulfillment/triple.py [hunk 3]
```diff
@@ -32,17 +32,34 @@
 _IN_ARROW = '<-'
 _OUT_TITLE = 'The {property} for {entity} is:'
 _IN_TITLE = '{entity} is the {property} for:'
+_MULTI_ENTITY_TITLE = 'The {property} for {entity} are as follows:'
+_MAX_ENTITIES_IN_TITLE = 2
+
+
+# Gets a title string for a list of entities
+def _get_entity_string(entities: List[Entity]) -> str:
+  if len(entities) == 1:
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply server/lib/nl/fulfillment/triple.py:46]: done

File: server/lib/nl/fulfillment/triple.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chejennifer) [reply server/lib/nl/fulfillment/triple.py]: updated the logic to allow for 3 entities in title and use `3 - 1` to get the number of entities in the list. I realized I was missing the logic for "and x more" when there are more than 3 entities, so in that case, I think it looks a little better not in a loop.. but take a look and let me know what you think

File: server/lib/nl/fulfillment/triple.py [hunk 4]
```diff
@@ -33,22 +33,36 @@
...
-  entity_str = ', '.join(
-      [e.name or e.dcid for e in entities[:entities_cutoff_idx]])
-  entity_str += f' and {entities[entities_cutoff_idx].name or entities[entities_cutoff_idx].dcid}'
+    return _entity_name(entities[0])
+  entity_str = ''
+  max_entries = min(_MAX_ENTITIES_IN_TITLE, len(entities))
+  # Join together the first max_entries - 1 number of entities with a ', '
+  entity_str = ', '.join([_entity_name(e) for e in entities[:max_entries - 1]])
+  if max_entries < len(entities):
+    # If not all entities are named in the title, get the number of unnamed
```
Comment (pradh) [inline server/lib/nl/fulfillment/triple.py:53]: In this if case, we will only print 2 and that's intentional?

**Review by pradh — APPROVED** [review]
Thanks for the edits.

File: server/lib/nl/fulfillment/triple.py (hunk 4 above)
Comment (chejennifer) [reply server/lib/nl/fulfillment/triple.py:53]: yes, that's intentional! From the mocks, we want 'The [property] for [entity1], [entity2] and [N more other] are as follows'

**Comment by chejennifer** [comment]
thanks for the review!

**Pull request merged**